byterun/caml/mlvalues.h

```diff
@@ -186,7 +183,24 @@ bits  63        (64-P) (63-P)        10 9     8 7   0
 /* Pointer to the first field. */
 #define Op_val(x) ((value *) (x))
 /* Fields are numbered from 0. */
+#if CAML_API_VERSION < 405
 #define Field(x, i) (((value *)(x)) [i])           /* Also an l-value. */
+#endif
+
+static inline void caml_read_field (value obj, mlsize_t f, value* ret)
+{
+  *ret = Op_val(obj)[f];
+}
+
+static inline value Field_imm (value obj, mlsize_t f)
+{
+  return Op_val(obj)[f];
+}
+
+#define Int_field(x, i) Int_val(Op_val(x)[i])
+#define Long_field(x, i) Long_val(Op_val(x)[i])
```

Comment:
On a common platform, Linux x86-64, the width of "int" is only 32 bits.  I've seen a surprising number of mistakes in OCaml FFI bindings where `Int_val` is used instead of `Long_val` and there is the potential for overflow.  I wonder if we could fix this somehow along with these other API changes.